test/kubernetes/e2e/features/services/tcproute/suite.go

```diff
@@ -31,95 +37,271 @@ func NewTestingSuite(ctx context.Context, testInst *e2e.TestInstallation) suite.
+			s.Require().NoError(err, fmt.Sprintf("Failed to apply manifest %s", tc.tcpRouteManifest))
+
+			// Assert gateway conditions
+			s.testInstallation.Assertions.EventuallyGatewayCondition(s.ctx, tc.gtwName, tc.gtwNs, v1.GatewayConditionAccepted, metav1.ConditionTrue, timeout)
+
+			// Set the expected status conditions based on the test case
+			expected := metav1.ConditionTrue
+			if tc.name == "CrossNamespaceTCPRouteWithoutReferenceGrant" {
+				expected = metav1.ConditionFalse
+			}
+
+			// Assert TCPRoute conditions
+			for _, tcpRouteName := range tc.tcpRouteNames {
+				s.testInstallation.Assertions.EventuallyTCPRouteCondition(s.ctx, tcpRouteName, tc.gtwNs, v1.RouteConditionAccepted, metav1.ConditionTrue, timeout)
+				s.testInstallation.Assertions.EventuallyTCPRouteCondition(s.ctx, tcpRouteName, tc.gtwNs, v1.RouteConditionResolvedRefs, expected, timeout)
+			}
+
+			// Assert gateway programmed condition
+			s.testInstallation.Assertions.EventuallyGatewayCondition(s.ctx, tc.gtwName, tc.gtwNs, v1.GatewayConditionProgrammed, metav1.ConditionTrue, timeout)
+
+			// Assert listener attached routes
+			for i, listenerName := range tc.listenerNames {
+				expectedRouteCount := tc.expectedRouteCounts[i]
+				s.testInstallation.Assertions.EventuallyGatewayListenerAttachedRoutes(s.ctx, tc.gtwName, tc.gtwNs, listenerName, expectedRouteCount, timeout)
+			}
+
+			// Assert expected responses
+			for i, port := range tc.ports {
+				if tc.name == "CrossNamespaceTCPRouteWithoutReferenceGrant" {
+					s.testInstallation.Assertions.AssertEventualCurlError(
+						s.ctx,
+						s.execOpts(tc.gtwNs),
+						[]curl.Option{
+							curl.WithHost(kubeutils.ServiceFQDN(tc.proxyService.ObjectMeta)),
+							curl.WithPort(port),
+							curl.VerboseOutput(),
+						},
+						tc.expectedErrorCode)
+				} else {
+					s.testInstallation.Assertions.AssertEventualCurlResponse(
+						s.ctx,
+						s.execOpts(tc.gtwNs),
+						[]curl.Option{
+							curl.WithHost(kubeutils.ServiceFQDN(tc.proxyService.ObjectMeta)),
+							curl.WithPort(port),
+							curl.VerboseOutput(),
+						},
+						tc.expectedResponses[i])
+				}
+			}
+		})
+	}
+}
+
+func validateManifestFile(path string) error {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return fmt.Errorf("Manifest file not found: %s", path)
+	}
+	return nil
+}
+
+func (s *testingSuite) setupTestEnvironment(nsManifest, gtwName, gtwNs, gtwManifest, svcManifest string, proxySvc *corev1.Service, proxyDeploy *appsv1.Deployment) {
+	s.applyManifests(nsManifest)
+
+	s.applyManifests(gtwManifest)
+	s.testInstallation.Assertions.EventuallyGatewayCondition(s.ctx, gtwName, gtwNs, v1.GatewayConditionAccepted, metav1.ConditionTrue, timeout)
+
+	s.applyManifests(svcManifest)
+	s.testInstallation.Assertions.EventuallyObjectsExist(s.ctx, proxySvc, proxyDeploy)
+}
+
+func (s *testingSuite) applyManifests(manifests ...string) {
+	for _, manifest := range manifests {
+		s.Eventually(func() bool {
```

Comment:
Why do we need to wrap the apply in an eventually? We've had issues in the past doing this (https://github.com/solo-io/solo-projects/issues/6321) where tests were masking incorrect product behavior, so I'd like to better understand why the eventually is necessary

Reply:
Resolved in commit https://github.com/solo-io/gloo/commit/281ace000966aa7f15d960ab73d60efc4fa1c189 by removing the `Eventually()` wrapper.